bigtable_rs/src/bigtable.rs

```diff
@@ -314,44 +304,69 @@ impl BigTableConnection {
-                    client: create_client(endpoints, auth_manager, is_read_only),
+                    client: create_client(channel, auth_manager, is_read_only),
                     table_prefix: Arc::new(table_prefix),
                     timeout: Arc::new(timeout),
                 })
             }
         }
     }
 
-    fn new_with_emulator(
+    /// Establish a connection to a BigTable emulator at [emulator_endpoint].
+    /// This is usually covered by [Self::new] or [Self::new_with_auth_manager],
+    /// which both support the `BIGTABLE_EMULATOR_HOST` env variable. However,
+    /// this function can also be used directly, in case setting
+    /// `BIGTABLE_EMULATOR_HOST` is inconvenient.
+    pub async fn new_with_emulator(
         emulator_endpoint: &str,
         project_id: &str,
         instance_name: &str,
         is_read_only: bool,
-        channel_size: usize,
         timeout: Option<Duration>,
     ) -> Result<Self> {
         info!("Connecting to bigtable emulator at {}", emulator_endpoint);
-        let endpoints: Vec<Endpoint> = vec![0; channel_size.max(1)]
-            .iter()
-            .map(move |_| {
-                Channel::from_shared(format!("http://{}", emulator_endpoint))
-                    .expect("Invalid connection emulator uri")
-                    .http2_keep_alive_interval(Duration::from_secs(60))
-                    .keep_alive_while_idle(true)
-            })
-            .map(|ep| {
-                if let Some(timeout) = timeout {
-                    ep.timeout(timeout)
-                } else {
-                    ep
-                }
-            })
-            .collect();
+
+        // Parse emulator_endpoint. Officially, it's only host:port,
+        // but unix:///path/to/unix.sock also works in the Go SDK at least.
+        // Having the emulator listen on unix domain sockets without ip2unix is
+        // covered in https://github.com/googleapis/google-cloud-go/pull/9665.
+
+        let endpoint = if emulator_endpoint.starts_with("unix://") {
+            // the URL doesn't matter, we use a custom connector later.
+            Endpoint::from_static("http://[::]:50051")
+        } else {
+            Channel::from_shared(format!("http://{}", emulator_endpoint))
+                .expect("invalid connection emulator uri")
+        };
+
+        let endpoint = endpoint
+            .http2_keep_alive_interval(Duration::from_secs(60))
+            .keep_alive_while_idle(true);
+
+        let endpoint = if let Some(timeout) = timeout {
+            endpoint.timeout(timeout)
+        } else {
+            endpoint
+        };
+
+        let channel = if let Some(path) = emulator_endpoint.strip_prefix("unix://") {
+            let path: String = path.to_string();
+            let connector = tower::service_fn({
+                move |_: tonic::transport::Uri| UnixStream::connect(path.clone())
+            });
+
+            endpoint.connect_with_connector(connector).await?
```

Comment:
Considering [`Channel::balance_list`](https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html#method.balance_list) is lazy, this should be lazy, too.

Reply:
Thanks for the update. I saw that you marked this as a draft, so perhaps you want to change more places?

What I think is that, as you mentioned as well, the `Channel::balance_list` is not async, so perhaps we can use that `endpoint.new_with_auth_manager_lazy()` here, so avoid calling an async function, so we can keep `new_with_auth_manager()` non-async as before (no breaking APIs).

What do you think?

Reply:
Oh, indeed I forgot to push my changes. I was testing them on a separate branch, but forgot to push them here once I finished.

In that version, `new_with_emulator` uses `endpoint.connect[_with_connector]_lazy`, so the whole function stays non async, causing the rest of the function to not have to change color.